dm/dm/master/scheduler/scheduler.go

```diff
@@ -2190,37 +2142,47 @@ func (s *Scheduler) handleWorkerOffline(ev ha.WorkerEvent, toLock bool) error {
 	}
 
 	// 2. find the bound relationship.
-	bound := w.Bound()
+	bounds := w.Bounds()
 
 	// 3. check whether bound before.
-	if bound.Source == "" {
+	if len(bounds) == 0 {
 		// 3.1. change the stage (from Free) to Offline.
 		w.ToOffline()
 		s.logger.Info("worker not bound, no need to unbound", zap.Stringer("event", ev))
 		return nil
 	}
 
 	// 4. delete the bound relationship in etcd.
-	_, err := ha.DeleteSourceBound(s.etcdCli, bound.Worker)
+	_, err := ha.DeleteSourceBoundByWorker(s.etcdCli, w.baseInfo.Name)
 	if err != nil {
 		return err
 	}
 
 	// 5. unbound for the source.
-	s.updateStatusToUnbound(bound.Source)
+	unbounds := make([]string, 0, len(bounds))
+	for _, bound := range bounds {
+		s.logger.Debug("unbound the worker for source", zap.Stringer("bound", bound), zap.Stringer("event", ev))
+		s.updateStatusToUnbound(bound.Source)
+		unbounds = append(unbounds, bound.Source)
+	}
 	defer func() {
```

Comment:
if we successful `tryBoundForSource` in step 7, lastBound should be the result of `tryBoundForSource`. Otherwise lastBound can be the offlined worker.

And I remember lastBound is maintained when bound, no need to be maintained when unbound.

Reply:
> if we successful tryBoundForSource in step 7, lastBound should be the result of tryBoundForSource. Otherwise lastBound can be the offlined worker.

If we didn't successfully find a worker for this source, its last bound worker is still this offline worker. We can try bound this worker after it's online again.

> And I remember lastBound is maintained when bound, no need to be maintained when unbound.

Because this PR has changed `lastBound` from `map[worker]source` to `map[source]worker`.

So we can no longer maintain this when we `tryBoundForSource` because this function doesn't know the last bound. However to avoid this affect bound decision, we update this after `tryBoundForSource` function.

Reply:
> If we didn't successfully find a worker for this source, its last bound worker is still this offline worker. We can try bound this worker after it's online again.

I agree with this. But the problem is the defer in line 2167 will always write `lastBound` to the offine worker, not the worker from `tryBoundForSource`

I think `lastBound` is also the current bound, so when worker offline we don't need to update it. If `lastBound` is the current bound, we don't need let `tryBoundForSource` know last bound, it just need to update it.

Reply:
To me, writing to last offline worker is more reasonable.

Imagine this situation:
1. source1 -> worker1
2. worker1 offline, source1 -> worker2
3. worker2 offline and worker1 online, if we want source1 bound back to worker1, `lastBound[source1]` should be worker1. Otherwise, it will treat worker1 the same as other workers.

Reply:
> worker2 offline and worker1 online

If the definition is "writing to last offline worker", here lastBound is worker2. Please give a better definition.

Reply:
addressed in https://github.com/pingcap/tiflow/pull/5077/commits/7baeaee4c364a60deee26ecb4660fb94f79b4296